providers/default_rhel.rb

```diff
@@ -0,0 +1,136 @@
+#
+# Cookbook Name:: kernel-modules
+# Author:: Jeremy MAURO <j.mauro@criteo.com>
+#
+# Copyright 2016, Criteo.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+provides :kernel_module, platform_family: 'rhel'
+
+MODPROBE_BIN = '/sbin/modprobe'.freeze
+
+def whyrun_supported?
+  true
+end
+
+use_inline_resources
+
+action :configure do
+  # Requirements
+  package node['kernel_modules']['packages']
+
+  # Module init loading section
+  template modload_file do
+    source "#{node['platform_family']}-#{node['platform_version'].to_i}/init-load.erb"
+    cookbook 'kernel-modules'
+    mode '0755'
+    variables(
+      name: new_resource.name,
+      bin: MODPROBE_BIN,
+    )
+    action new_resource.onboot ? :create : :delete
+  end
+
+  # Module loading options
+  template modprobe_file do
+    source 'modprobe.conf.erb'
+    cookbook 'kernel-modules'
+    mode '0644'
+    variables(
+      name: new_resource.name,
+      alias: new_resource.alias,
+      options: new_resource.options,
+      install: new_resource.install,
+      remove: new_resource.remove,
+      blacklist: new_resource.blacklist,
+    )
+    action :create
+  end
+end
+
+action :load do
+  # When loading a module it means the module must be configured
+  run_action :configure
+
+  wanted_action = blacklisted?(new_resource.install, new_resource.blacklist) ? :run : :nothing
+  execute "Unloading module #{new_resource.name}" do
+    command reload_cmd
+    only_if "lsmod | grep -q #{new_resource.name}"
```

Comment:
Everything would be clearer by leveraging custom resource methods and `load_current_resource`.

For example we could have `current_resource.loaded?` or `new_resource.blacklisted? `

Then we should avoid calling the unload command as "reload_cmd".

Finally the subscription is broken it will only unload the module. I suggest to have one execute resource in the configure action that is with `action :nothing` and `only_if { current_resource.loaded? }` and a notification from the templat set up if reload is true.

Then here we only worry about executing a load in the load action.